### PR TITLE
bpo-23819: Fix asyncio tests on python optimized mode

### PR DESCRIPTION
**Pull request by kumaraditya303**
https://bugs.python.org/issue23819

### CODE REVIEW

**Comment by kumaraditya303** [comment]
cc @asvetlov

File: Lib/asyncio/base_events.py [hunk 1]
```diff
@@ -1285,9 +1285,8 @@ async def create_datagram_endpoint(self, protocol_factory,
                 addr_infos = {}  # Using order preserving dict
                 for idx, addr in ((0, local_addr), (1, remote_addr)):
                     if addr is not None:
-                        assert isinstance(addr, tuple) and len(addr) == 2, (
-                            '2-tuple is expected')
-
+                        if not (isinstance(addr, tuple) and len(addr) == 2):
+                            raise ValueError(f'addr should be a tuple of length 2 not {addr!r}')
```
Comment (serhiy-storchaka) [inline Lib/asyncio/base_events.py]: There is no addr parameter. There are local_addr and remote_addr. Depending on family they can be None, str, etc, not only 2-tuples.

I think that TypeError is more appropriate here.

File: Lib/asyncio/proactor_events.py [hunk 2]
```diff
@@ -274,8 +274,12 @@ def _loop_reading(self, fut=None):
         data = None
         try:
             if fut is not None:
-                assert self._read_fut is fut or (self._read_fut is None and
-                                                 self._closing)
+                if not (self._read_fut is fut or (self._read_fut is None and
+                                             self._closing)):
+                    raise RuntimeError(
+                            f'_loop_reading called from unexpected future: '
+                            f'{self._read_fut} != {fut}')
```
Comment (serhiy-storchaka) [inline Lib/asyncio/proactor_events.py]: Is it even called with fut is not None?

**Review by asvetlov — CHANGES_REQUESTED** [review]
Thanks for the PR!
I agree, *public faced* assertions should be converted to explicit errors.
But internal invariant checks like `assert self._buffer` in `_write_ready()` method should be untouched.
If this particular assertion breaks tests with `-O` mode -- the check should be skipped if `sys.flags.optimize` is set.

**Comment by kumaraditya303** [comment]
I have made the requested changes; please review again

File: Lib/test/test_asyncio/test_base_events.py [hunk 3]
```diff
@@ -1590,7 +1590,8 @@ def test_create_datagram_endpoint_no_addrinfo(self, m_socket):
             MyDatagramProto, local_addr=('localhost', 0))
         self.assertRaises(
             OSError, self.loop.run_until_complete, coro)
-
+            
+    @unittest.skipIf(sys.flags.optimize, "Assertions are disabled in optimized mode")
```
Comment (serhiy-storchaka) [inline Lib/test/test_asyncio/test_base_events.py]: It is not correct. The code should raise a TypeError. I think a simple `raise TypeError('2-tuple is expected')` should be enough, there are precedents in the same function. 

Other tests can be skipped, they test internal methods.

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_base_events.py]: fixed!

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/test/test_asyncio/test_base_events.py [hunk 4]
```diff
@@ -1590,16 +1590,16 @@ def test_create_datagram_endpoint_no_addrinfo(self, m_socket):
             MyDatagramProto, local_addr=('localhost', 0))
         self.assertRaises(
             OSError, self.loop.run_until_complete, coro)
-
+
```
Comment (serhiy-storchaka) [inline Lib/test/test_asyncio/test_base_events.py]:
```suggestion

```

**Comment by kumaraditya303** [comment]
@asvetlov I have addressed your comments.

**Pull request merged**